components/salsa-2022-macros/src/input.rs

```diff
@@ -112,8 +104,16 @@ impl InputStruct {
         )
         .collect();
 
+        // setters
         let set_field_names = self.all_set_field_names();
```

Comment:
What's missing here is that we need to not generate a "setter" for an `#[id]` field (and we should make a negative test that shows no setter exists). The idea is that you cannot change the value of an `#[id]` field after `new` is invoked, and that's why they are safe to include in the debug impl.

It occurs to me that, on singleton inputs, with #415, their fields can be modified when a second instance is created, so we probably want to prohibit `#[id]` fields if the type is marked as singleton. This pushes me further towards the option (described in that issue) of panicking when a second instance is created.

Reply:
One way to achieve this goal would be to modify

https://github.com/salsa-rs/salsa/blob/32f0ce0cffc0e3695a3292fb740d378cdcf19408/components/salsa-2022-macros/src/salsa_struct.rs#L124-L127

so that  it doesn't include fields with the `#[id]` attribute. Also, we may want to move that function out of `salsa_struct.rs` and into `input.rs`, I think only inputs have setters, so it's not really common code that applies to all kinds of salsa structs.